### PR TITLE
feat: add replay

### PR DESCRIPTION
**Pull request by arinda-arif**

### CODE REVIEW

File: core/scheduler/service/replay_manager.go [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+	if err != nil {
+		m.l.Error("unable to get on going replay")
+	}
+
+	for _, replay := range onGoingReplays {
+		runningTime := m.Now().Sub(replay.CreatedAt())
+		if runningTime < m.config.ReplayTimeout {
+			continue
+		}
+		message := fmt.Sprintf("replay is timing out. %s", replay.Message())
```
Comment (sravankorumilli) [inline core/scheduler/service/replay_manager.go:91]: is there a scenario where replay message is updated twice if there is it is better to prefix it. & we could rename the message as "replay timed out"

File: core/scheduler/service/replay_manager.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+		if err := m.replayRepository.UpdateReplayStatus(ctx, replay.ID(), scheduler.ReplayStateFailed, message); err != nil {
+			m.l.Error("unable to mark replay %s as failed due to time out", replay.ID())
+		}
+		m.l.Info("replay %s is timing out. marked as failed.", replay.ID())
```
Comment (sravankorumilli) [inline core/scheduler/service/replay_manager.go:95]: similar replay timed out.

File: core/scheduler/service/replay_service.go [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+	validator ReplayValidator
+}
+
+func (r ReplayService) CreateReplay(ctx context.Context, tenant tenant.Tenant, jobName scheduler.JobName, config *scheduler.ReplayConfig) (replayID uuid.UUID, err error) {
+	subjectJob, err := r.jobRepo.GetJobDetails(ctx, tenant.ProjectName(), jobName)
+	if err != nil {
+		return uuid.Nil, fmt.Errorf("unable to get job details from DB for jobName: %s, project:%s,  error:%w ", jobName, tenant.ProjectName().String(), err)
+	}
+
+	jobCron, err := cron.ParseCronSchedule(subjectJob.Schedule.Interval)
```
Comment (sravankorumilli) [inline core/scheduler/service/replay_service.go:40]: in which scenario job is add but cron can't be parsed

Comment (sravankorumilli) [reply core/scheduler/service/replay_service.go:40]: if this is not expected just add explicitly this shouldn't happen in a realistic scenario

File: internal/store/postgres/migrations/000050_replace_replay.up.sql [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    status  VARCHAR(30) NOT NULL,
+    message TEXT,
+
+    created_at TIMESTAMP WITH TIME ZONE NOT NULL,
+    updated_at TIMESTAMP WITH TIME ZONE NOT NULL
+);
+
+CREATE INDEX IF NOT EXISTS replay_request_project_name_idx on replay_request(project_name);
+CREATE INDEX IF NOT EXISTS replay_request_job_name_idx on replay_request(job_name);
+CREATE INDEX IF NOT EXISTS replay_request_start_time_idx on replay_request(start_time);
```
Comment (sravankorumilli) [inline internal/store/postgres/migrations/000050_replace_replay.up.sql:28]: we don't need index on start & end_time just job_name should be sufficient.

Comment (sravankorumilli) [reply internal/store/postgres/migrations/000050_replace_replay.up.sql:28]: as the table is very small we don't need to look into optimizing by creating many indexes

**Review by sravankorumilli — CHANGES_REQUESTED** [review]

File: config/config_server.go [hunk 5]
```diff
@@ -53,3 +56,10 @@ type ResourceManagerConfigOptimus struct {
 type PluginConfig struct {
 	Artifacts []string `mapstructure:"artifacts"`
 }
+
+// TODO: add worker interval
+type ReplayConfig struct {
+	WorkerTimeout time.Duration `mapstructure:"worker_timeout" default:"10m"`
```
[the remark this answers is not in the document]
Comment (arinda-arif) [reply config/config_server.go:62]: agreed to not configure it.

File: config/config_server.go [hunk 6]
```diff
@@ -53,3 +56,10 @@ type ResourceManagerConfigOptimus struct {
...
+	ReplayTimeout time.Duration `mapstructure:"replay_timeout" default:"3h"`
+	LoopTimeout   time.Duration `mapstructure:"replay_timeout" default:"5m"`
```
[the remark this answers is not in the document]
Comment (arinda-arif) [reply config/config_server.go:64]: agreed to not configure it.

File: client/cmd/replay/create.go [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+
+	"github.com/odpf/optimus/client/cmd/internal"
+	"github.com/odpf/optimus/client/cmd/internal/connectivity"
+	"github.com/odpf/optimus/client/cmd/internal/logger"
+	"github.com/odpf/optimus/config"
+	pb "github.com/odpf/optimus/protos/odpf/optimus/core/v1beta1"
+)
+
+const (
+	replayTimeout = time.Minute * 60
```
[the remark this answers is not in the document]
Comment (arinda-arif) [reply client/cmd/replay/create.go:21]: reduce the number, 1 minute should be enough

File: core/scheduler/service/replay_manager.go (hunk 1 above)
Comment (arinda-arif) [reply core/scheduler/service/replay_manager.go:91]: message should not be updated twice.

Comment (arinda-arif) [reply core/scheduler/service/replay_manager.go:91]: will change to `timed out`.

File: core/scheduler/service/replay_service.go (hunk 3 above)
Comment (arinda-arif) [reply core/scheduler/service/replay_service.go:40]: in the error message, should mention this is not the expected scenario.

**Pull request merged**